### PR TITLE
refactor: Remove SQL triggers in favor of plain old GO code

### PR DESCRIPTION
**Pull request by gfyrag**
# Refactor: Remove SQL triggers in favor of plain old GO code.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Refactoring / Technical debt

## What parts of the code are impacted ?
Most on the storage, some parts of the core ledger.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pkg/storage/sqlstorage/transactions.go [hunk 1]
```diff
@@ -267,3 +269,171 @@
...
+		return err
+	}
+	ub.
+		Update(s.schema.Table("transactions")).
+		Where(ub.E("id", id))
+
+	switch Flavor(s.schema.Flavor()) {
+	case PostgreSQL:
+		ub.Set(fmt.Sprintf("metadata = metadata || '%s'::jsonb", string(metadataData)))
+	case SQLite:
```
[the remark this answers is not in the document]
Comment (gfyrag) [reply pkg/storage/sqlstorage/transactions.go:390]: Was modified but github seems not update the check.

File: pkg/core/metadata.go [hunk 2]
```diff
@@ -19,8 +19,36 @@ func SpecMetadata(name string) string {
 	return numaryNamespace + name
 }
 
+type RawMetadata map[string]any
```
Comment (antoinegelloz) [inline pkg/core/metadata.go]: I don't understand why we need a new struct here? I feel like we should try to use the same `Metadata`.
Why not change it to `type Metadata map[string]any`?

File: pkg/core/transaction.go [hunk 3]
```diff
@@ -34,13 +34,54 @@ func (t *TransactionData) Reverse() TransactionData {
 
 var _ json.Marshaler = Transaction{}
```
Comment (antoinegelloz) [inline pkg/core/transaction.go]: What is it for?

File: pkg/core/transaction.go [hunk 4]
```diff
@@ -34,13 +34,54 @@ func (t *TransactionData) Reverse() TransactionData {
...
 
+type RawTransaction struct {
```
Comment (antoinegelloz) [inline pkg/core/transaction.go]: I also feel like we should avoid creating this new struct. Why do you need it?

File: pkg/ledger/ledger.go [hunk 5]
```diff
@@ -218,18 +219,16 @@ func (l *Ledger) SaveMeta(ctx context.Context, targetType string, targetID inter
...
-	if err != nil {
-		return err
-	}
-
-	log := core.NewSetMetadataLog(lastLog, core.SetMetadata{
-		TargetType: strings.ToUpper(targetType),
-		TargetID:   targetID,
-		Metadata:   m,
+	err = l.store.WithTX(ctx, func(api storage.API) error {
+		switch targetType {
```
Comment (antoinegelloz) [inline pkg/ledger/ledger.go:224]: The `if targetType != core.MetaTargetTypeTransaction && targetType != core.MetaTargetTypeAccount` above could be merged with this switch to check the type only once

File: pkg/ledger/ledger_test.go [hunk 6]
```diff
@@ -437,11 +437,11 @@ func TestTransactionMetadata(t *testing.T) {
 		require.NoError(t, err)
 
 		meta, ok := tx.Metadata["a random metadata"]
-		assert.True(t, ok)
+		require.True(t, ok)
```
Comment (antoinegelloz) [inline pkg/ledger/ledger_test.go:459]: I think that using `assert` here was more appropriate, as this is what we want to test. I like to separate error checking with `require` for things that are useful for the test but isn't what we are testing, and `assert` for things that we are actually testing. What do you think?

File: pkg/opentelemetry/opentelemetrytraces/storage_test.go [hunk 7]
```diff
@@ -72,7 +72,7 @@ func TestStore(t *testing.T) {
 }
 
 func testAppendLog(t *testing.T, store storage.Store) {
-	err := store.AppendLog(context.Background(), core.NewTransactionLog(nil, core.Transaction{}))
+	err := store.Commit(context.Background(), core.Transaction{})
```
Comment (antoinegelloz) [inline pkg/opentelemetry/opentelemetrytraces/storage_test.go]: This test could be deleted, doesn't make sense anymore

File: pkg/storage/sqlstorage/accounts.go [hunk 8]
```diff
@@ -197,6 +194,42 @@ func (s *Store) getAccount(ctx context.Context, exec executor, addr string) (*co
...
+		Cols("address", "metadata").
+		Values(account, "{}").
+		SQL("ON CONFLICT DO NOTHING").
+		BuildWithFlavor(s.schema.Flavor())
+
+	_, err := s.executor.ExecContext(ctx, sqlq, args...)
+	return s.error(err)
+}
+
+func (s *API) UpdateAccountMetadata(ctx context.Context, address string, metadata core.Metadata, at time.Time) error {
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go:211]: Warning: `at` arg is unused

File: pkg/storage/sqlstorage/commit.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	if err := s.insertTransactions(ctx, txs...); err != nil {
+		return nil, errors.Wrap(err, "inserting transactions")
+	}
+
+	postCommitVolumes := core.AggregatePostCommitVolumes(txs...)
+
+	for account := range postCommitVolumes {
+		err := s.ensureAccountExists(ctx, account)
+		if err != nil {
+			return nil, errors.Wrap(err, "creating account entry")
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/commit.go]:
```suggestion
			return nil, errors.Wrap(err, "ensuring account exists")
```
?

File: pkg/storage/sqlstorage/migrations_test.go [hunk 10]
```diff
@@ -259,58 +258,42 @@ var postMigrate = map[string]func(t *testing.T, store *sqlstorage.Store){
...
-		if !assert.NoError(t, err) {
-			return
-		}
-		if !assert.Len(t, logs, 8) {
-			return
-		}
+		require.NoError(t, err)
+		require.Len(t, logs, 8)
 
+		fmt.Println("check all hash")
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/migrations_test.go]:
```suggestion
```

File: pkg/storage/sqlstorage/migrations_test.go [hunk 11]
```diff
@@ -155,8 +155,7 @@ var postMigrate = map[string]func(t *testing.T, store *sqlstorage.Store){
 			return
 		}
 	},
-	"1": func(t *testing.T, store *sqlstorage.Store) {},
-	"8": func(t *testing.T, store *sqlstorage.Store) {
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/migrations_test.go:159]: Why removing previous migration tests?

File: pkg/storage/sqlstorage/migrations_test.go [hunk 12]
```diff
@@ -155,8 +155,7 @@ var postMigrate = map[string]func(t *testing.T, store *sqlstorage.Store){
...
+	"10": func(t *testing.T, store *sqlstorage.Store) {
 
 		count, err := store.CountTransactions(context.Background(), storage.TransactionsQuery{})
 		if !assert.NoError(t, err) {
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/migrations_test.go]: I think that a cleanup of this file is necessary to remove this kind of unnecessary lines of code

File: pkg/storage/sqlstorage/store.go [hunk 13]
```diff
@@ -12,33 +13,41 @@ const (
...
+		executor: executor,
+	}
 }
 
-func (s *Store) Name() string {
+func (s *API) Name() string {
 	return s.schema.Name()
 }
 
+var _ storage.API = &API{}
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/store.go:43]: What is this line for?

File: pkg/storage/sqlstorage/store.go [hunk 14]
```diff
@@ -54,4 +63,33 @@ func (s *Store) Close(ctx context.Context) error {
...
+
+func NewStore(schema Schema, onClose func(ctx context.Context) error) *Store {
+	return &Store{
+		API:     NewAPI(schema, schema),
+		schema:  schema,
+		onClose: onClose,
+	}
+}
+
 var _ storage.Store = &Store{}
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/store.go:95]: And this one?

File: pkg/storage/storage.go [hunk 15]
```diff
@@ -70,24 +71,45 @@ type Store interface {
...
 	GetBalancesAggregated(context.Context, BalancesQuery) (core.AssetsBalances, error)
-
-	AppendLog(ctx context.Context, log ...core.Log) error
 	LastLog(ctx context.Context) (*core.Log, error)
 	Logs(ctx context.Context) ([]core.Log, error)
-
 	LoadMapping(ctx context.Context) (*core.Mapping, error)
+
+	UpdateTransactionMetadata(ctx context.Context, id uint64, metadata core.Metadata, at time.Time) error
+	UpdateAccountMetadata(ctx context.Context, id string, metadata core.Metadata, at time.Time) error
```
Comment (antoinegelloz) [inline pkg/storage/storage.go:81]: Unused `at time.Time` arg

File: pkg/storage/storage.go [hunk 16]
```diff
@@ -60,7 +61,7 @@ func IsTooManyClientError(err error) bool {
 	return IsErrorCode(err, TooManyClient)
 }
 
-type Store interface {
+type API interface {
```
Comment (antoinegelloz) [inline pkg/storage/storage.go:64]: Why did you need to add this interface?
I find the naming confusing, as this interface is a collection of storage methods.
`APIStore` maybe?

File: pkg/storage/storage.go [hunk 17]
```diff
@@ -70,24 +71,45 @@ type Store interface {
...
 	SaveMapping(ctx context.Context, m core.Mapping) error
-	Initialize(context.Context) (bool, error)
 	Name() string
+}
+
+type Store interface {
+	API
+	Initialize(context.Context) (bool, error)
 	Close(context.Context) error
+	WithTX(ctx context.Context, callback func(api API) error) error
```
Comment (antoinegelloz) [inline pkg/storage/storage.go:91]: I can't figure out what this method is doing 😅

File: pkg/storage/sqlstorage/store_bench_test.go [hunk 18]
```diff
@@ -90,10 +90,10 @@ func BenchmarkStore(b *testing.B) {
 					return
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/store_bench_test.go]: I think it's time to remove all of those benchmarks, they don't make sense anymore and are not used.
What do you think?

File: pkg/storage/sqlstorage/store.go [hunk 19]
```diff
@@ -54,4 +63,33 @@ func (s *Store) Close(ctx context.Context) error {
...
+
+func (s *Store) WithTX(ctx context.Context, callback func(api storage.API) error) error {
+	return s.withSqlTx(ctx, func(tx *sql.Tx) error {
+		return callback(NewAPI(s.schema, tx))
+	})
+}
+
+func NewStore(schema Schema, onClose func(ctx context.Context) error) *Store {
+	return &Store{
+		API:     NewAPI(schema, schema),
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/store.go:89]: This is weird 😅

File: pkg/core/transaction.go (hunk 3 above)
Comment (gfyrag) [reply pkg/core/transaction.go]: This is a compile time assertion that struct Transaction implement json.Marshaler interface.

File: pkg/core/transaction.go (hunk 4 above)
Comment (gfyrag) [reply pkg/core/transaction.go]: The "RawTransaction" is the struct used to store transactions at logs level.
Basically, this struct contains fields required to rebuild the model reading only the logs.
So, it excludes pre/post commit volumes, and potentially, futures fields we will add.

File: pkg/ledger/ledger.go (hunk 5 above)
Comment (gfyrag) [reply pkg/ledger/ledger.go:224]: WithTx start a new sql transaction.
We should ensure all validations checks are ok before starting it as it consume a connection to the database.

File: pkg/ledger/ledger_test.go (hunk 6 above)
Comment (gfyrag) [reply pkg/ledger/ledger_test.go:459]: I prefer require in all cases :D
I don't care about continue to test if the check fail.
It also complicate the testing process as we need to think about the legitimacy of the assert vs the require.
It is a mental burden for a minimal gain.
But i can change my mind.

File: pkg/opentelemetry/opentelemetrytraces/storage_test.go (hunk 7 above)
Comment (gfyrag) [reply pkg/opentelemetry/opentelemetrytraces/storage_test.go]: Yes

File: pkg/storage/sqlstorage/migrations_test.go (hunk 11 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/migrations_test.go:159]: The "1" is empty, so useless, and "8" was moved to "10" as we just need to check the state after all migrations are done.
This test is bloated (TestAllMigrations), i think we should remove it and prefer dedicated for each migration.

File: pkg/storage/sqlstorage/migrations_test.go (hunk 12 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/migrations_test.go]: Completely agree

File: pkg/storage/sqlstorage/migrations_test.go (hunk 10 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/migrations_test.go]: Oups...

File: pkg/storage/sqlstorage/store.go (hunk 13 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/store.go:43]: Again, a compile time assertion to ensure *API implements storage.API interface.
Not really required in this case since the code will not compile if the structure does not implements the interface.
In the case of the json.Marshaler, it would compile.

Even if not strictly required in this case, it helps to see directy than *API implement storage.API.

Also, Intellij/Goland will propose you to implements the methods automatically ;)

File: pkg/storage/sqlstorage/store.go (hunk 14 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/store.go:95]: Same

File: pkg/storage/sqlstorage/store.go (hunk 19 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/store.go:89]: Yes, it works, maybe it will need a bit of rework.
The purpose of this PR is to remove triggers, we have to ensure it is working as expected, not lost too many time on implementation details.

File: pkg/storage/sqlstorage/store_bench_test.go (hunk 18 above)
Comment (gfyrag) [reply pkg/storage/sqlstorage/store_bench_test.go]: Yes, i'm agree.

File: pkg/storage/storage.go (hunk 16 above)
Comment (gfyrag) [reply pkg/storage/storage.go:64]: The API interface is just to have an interface without Close() and other methods available on the store, as those doesn't make sense on sql transaction context.

Comment (gfyrag) [reply pkg/storage/storage.go:64]: Can be renamed as well.

File: pkg/storage/storage.go (hunk 17 above)
Comment (gfyrag) [reply pkg/storage/storage.go:91]: Create a new transaction and run a callback.
The API interface embed automatically the transaction.

File: pkg/core/transaction.go (hunk 4 above)
Comment (antoinegelloz) [reply pkg/core/transaction.go]: We decided together to try embedding a new struct without volumes in the `Transaction` struct.

File: pkg/ledger/ledger.go (hunk 5 above)
Comment (antoinegelloz) [reply pkg/ledger/ledger.go:224]: Alright!
We said together that this `if` should be replaced by a `switch` to improve readability

File: pkg/storage/sqlstorage/migrations_test.go (hunk 11 above)
Comment (antoinegelloz) [reply pkg/storage/sqlstorage/migrations_test.go:159]: Ok let's remove it then :D

File: pkg/storage/sqlstorage/migrations_test.go (hunk 12 above)
Comment (antoinegelloz) [reply pkg/storage/sqlstorage/migrations_test.go]: Deleting this file is easier haha

File: pkg/storage/sqlstorage/store.go (hunk 13 above)
Comment (antoinegelloz) [reply pkg/storage/sqlstorage/store.go:43]: Alright thanks :)

File: pkg/storage/sqlstorage/store.go (hunk 19 above)
Comment (antoinegelloz) [reply pkg/storage/sqlstorage/store.go:89]: Alright :)

File: pkg/core/transaction.go [hunk 20]
```diff
@@ -32,58 +32,21 @@ func (t *TransactionData) Reverse() TransactionData {
...
-		Postings:  t.Postings,
-		Reference: t.Reference,
-		Metadata:  metadata,
-		ID:        t.ID,
-		Timestamp: t.Timestamp,
-	}
+type ExpandedTransaction struct {
+	Transaction
+	PreCommitVolumes  AccountsAssetsVolumes `json:"preCommitVolumes,omitempty"`  // Keep omitempty to keep consistent hash
+	PostCommitVolumes AccountsAssetsVolumes `json:"postCommitVolumes,omitempty"` // Keep omitempty to keep consistent hash
```
Comment (antoinegelloz) [inline pkg/core/transaction.go]:
```suggestion
	PreCommitVolumes  AccountsAssetsVolumes `json:"preCommitVolumes,omitempty"`
	PostCommitVolumes AccountsAssetsVolumes `json:"postCommitVolumes,omitempty"`
```
Not anymore ;)

File: pkg/ledger/executor_test.go [hunk 21]
```diff
@@ -376,6 +377,7 @@ func TestSetTxMeta(t *testing.T) {
 					last, err := l.store.GetLastTransaction(context.Background())
 					require.NoError(t, err)
 
+					spew.Dump(last.Metadata, tc.expectedMetadata)
```
Comment (antoinegelloz) [inline pkg/ledger/executor_test.go]:
```suggestion
```
Oupsy!

File: pkg/storage/sqlstorage/accounts.go [hunk 22]
```diff
@@ -230,6 +230,16 @@ func (s *API) UpdateAccountMetadata(ctx context.Context, address string, metadat
 
 	sqlq, args := ib.BuildWithFlavor(s.schema.Flavor())
 	_, err = s.executor.ExecContext(ctx, sqlq, args...)
+	if err != nil { return err }
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go]: File is not `gofmt`-ed
Weird the CI didn't run...

File: pkg/storage/sqlstorage/accounts.go [hunk 23]
```diff
@@ -230,6 +230,16 @@ func (s *API) UpdateAccountMetadata(ctx context.Context, address string, metadat
...
+	if err != nil {
+		return errors.Wrap(err, "reading last log")
+	}
 
-	return err
+	return s.appendLog(ctx, core.NewSetMetadataLog(lastLog, at, core.SetMetadata{
+		TargetType: core.MetaTargetTypeAccount,
+		TargetID:   address,
+		Metadata:   metadata,
+	}))
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go]: Oh wow didn't see this missing!

**Review by antoinegelloz — APPROVED** [review]

**Review by jdupas22 — APPROVED** [review]
